File: sdk/wauth-ts/src/rp.ts
import type { JSONWebKeySet } from "jose";

import type { JsonValue } from "./types.js";
import { verifyJwtWithJwks } from "./jwt.js";

export interface CapabilityValidationInput {
  claims: Record<string, JsonValue>;
  expectedAudience: string;
  expectedActionHash: string;
  nowEpochSeconds?: number;
  maxClockSkewSeconds?: number;
}

export interface CapabilityValidationResult {
  ok: boolean;
  errors: string[];
  replayKey?: string;
}

export interface CapabilityJwtVerificationInput {
  token: string;
  jwks: JSONWebKeySet;
  expectedIssuer?: string;
  expectedAudience: string;
  expectedActionHash: string;
  nowEpochSeconds?: number;
  maxClockSkewSeconds?: number;
  allowedAlgorithms?: string[];
}

export interface CapabilityJwtVerificationResult extends CapabilityValidationResult {
  claims?: Record<string, JsonValue>;
  header?: Record<string, JsonValue>;
}

function audienceIncludes(audClaim: JsonValue | undefined, expectedAudience: string): boolean {
  if (typeof audClaim === "string") {
    return audClaim === expectedAudience;
  }
  if (Array.isArray(audClaim)) {
    return audClaim.some((value) => value === expectedAudience);
  }
  return false;
}

export function validateCapabilityClaims(input: CapabilityValidationInput): CapabilityValidationResult {
  const now = input.nowEpochSeconds ?? Math.floor(Date.now() / 1000);
  const skew = input.maxClockSkewSeconds ?? 120;
  const errors: string[] = [];
  const claims = input.claims;

  if (!audienceIncludes(claims.aud, input.expectedAudience)) {
    errors.push("audience mismatch");
  }

  if (typeof claims.exp !== "number" || claims.exp < now - skew) {
    errors.push("token expired");
  }

  if (typeof claims.iat !== "number" || claims.iat > now + skew) {
    errors.push("issued-at is in the future beyond skew");
  }

  if (claims.action_hash !== input.expectedActionHash) {
    errors.push("action hash mismatch");
  }

  if (typeof claims.jti !== "string" || claims.jti.length === 0) {
    errors.push("missing jti for replay protection");
  }

  return {
    ok: errors.length === 0,
    errors,
    replayKey: typeof claims.jti === "string" ? claims.jti : undefined
  };
}

export async function verifyCapabilityJwtWithJwks(
  input: CapabilityJwtVerificationInput
): Promise<CapabilityJwtVerificationResult> {
  const jwtResult = await verifyJwtWithJwks({
    token: input.token,
    jwks: input.jwks,
    expectedIssuer: input.expectedIssuer,
    expectedAudience: input.expectedAudience,
    expectedActionHash: input.expectedActionHash,
    nowEpochSeconds: input.nowEpochSeconds,
    clockToleranceSeconds: input.maxClockSkewSeconds,
    allowedAlgorithms: input.allowedAlgorithms,
    requireJti: true
  });

  if (!jwtResult.ok || !jwtResult.claims) {
    return {
      ok: false,
      errors: jwtResult.errors,
      replayKey: jwtResult.replayKey,
      claims: jwtResult.claims,
      header: jwtResult.header as unknown as Record<string, JsonValue>
    };
  }

  const claimResult = validateCapabilityClaims({
    claims: jwtResult.claims,
    expectedAudience: input.expectedAudience,
    expectedActionHash: input.expectedActionHash,
    nowEpochSeconds: input.nowEpochSeconds,
    maxClockSkewSeconds: input.maxClockSkewSeconds
  });

  return {
    ok: jwtResult.ok && claimResult.ok,
    errors: [...jwtResult.errors, ...claimResult.errors],
    replayKey: claimResult.replayKey ?? jwtResult.replayKey,
    claims: jwtResult.claims,
    header: jwtResult.header as unknown as Record<string, JsonValue>
  };
}

File: sdk/wauth-py/src/wauth_sdk/rp.py
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Sequence

from .jwt_verify import verify_jwt_with_jwks


@dataclass
class CapabilityValidationResult:
    ok: bool
    errors: List[str]
    replay_key: Optional[str]


@dataclass
class CapabilityJwtVerificationResult(CapabilityValidationResult):
    claims: Optional[Dict[str, Any]]
    header: Optional[Dict[str, Any]]


def _audience_includes(aud_claim: Any, expected_audience: str) -> bool:
    if isinstance(aud_claim, str):
        return aud_claim == expected_audience
    if isinstance(aud_claim, list):
        return any(v == expected_audience for v in aud_claim)
    return False


def _extract_action_hash(claims: Dict[str, Any]) -> Optional[str]:
    action_hash = claims.get("action_hash")
    if isinstance(action_hash, str):
        return action_hash

    authorization_details = claims.get("authorization_details")
    if not isinstance(authorization_details, list):
        return None

    for detail in authorization_details:
        if isinstance(detail, dict) and isinstance(detail.get("action_hash"), str):
            return detail["action_hash"]

    return None


def validate_capability_claims(
    claims: Dict[str, Any],
    expected_audience: str,
    expected_action_hash: str,
    now_epoch_seconds: int,
    max_clock_skew_seconds: int = 120,
) -> CapabilityValidationResult:
    errors: List[str] = []

    if not _audience_includes(claims.get("aud"), expected_audience):
        errors.append("audience mismatch")

    exp = claims.get("exp")
    if not isinstance(exp, (int, float)) or exp < (now_epoch_seconds - max_clock_skew_seconds):
        errors.append("token expired")

    iat = claims.get("iat")
    if not isinstance(iat, (int, float)) or iat > (now_epoch_seconds + max_clock_skew_seconds):
        errors.append("issued-at is in the future beyond skew")

    action_hash = _extract_action_hash(claims)
    if action_hash != expected_action_hash:
        errors.append("action hash mismatch")

    jti = claims.get("jti")
    if not isinstance(jti, str) or len(jti) == 0:
        errors.append("missing jti for replay protection")

    return CapabilityValidationResult(
        ok=len(errors) == 0,
        errors=errors,
        replay_key=jti if isinstance(jti, str) else None,
    )


def verify_capability_jwt_with_jwks(
    token: str,
    jwks: Dict[str, Any],
    expected_audience: str,
    expected_action_hash: str,
    expected_issuer: Optional[str] = None,
    now_epoch_seconds: Optional[int] = None,
    max_clock_skew_seconds: int = 120,
    allowed_algorithms: Optional[Sequence[str]] = None,
) -> CapabilityJwtVerificationResult:
    jwt_result = verify_jwt_with_jwks(
        token=token,
        jwks=jwks,
        expected_issuer=expected_issuer,
        expected_audience=expected_audience,
        now_epoch_seconds=now_epoch_seconds,
        clock_tolerance_seconds=max_clock_skew_seconds,
        allowed_algorithms=allowed_algorithms,
        expected_action_hash=expected_action_hash,
        require_jti=True,
    )

    if not jwt_result.ok or jwt_result.claims is None:
        return CapabilityJwtVerificationResult(
            ok=False,
            errors=jwt_result.errors,
            replay_key=jwt_result.replay_key,
            claims=jwt_result.claims,
            header=jwt_result.header,
        )

    now = now_epoch_seconds if now_epoch_seconds is not None else 0
    if now == 0:
        # keep deterministic runtime behavior delegated to verify_jwt_with_jwks when no explicit clock is passed
        import time as _time

        now = int(_time.time())

    claims_result = validate_capability_claims(
        claims=jwt_result.claims,
        expected_audience=expected_audience,
        expected_action_hash=expected_action_hash,
        now_epoch_seconds=now,
        max_clock_skew_seconds=max_clock_skew_seconds,
    )

    return CapabilityJwtVerificationResult(
        ok=jwt_result.ok and claims_result.ok,
        errors=[*jwt_result.errors, *claims_result.errors],
        replay_key=claims_result.replay_key or jwt_result.replay_key,
        claims=jwt_result.claims,
        header=jwt_result.header,
    )
